lisp/ess-custom.el

```diff
@@ -2742,82 +2731,82 @@ for an example.")
   '((ess-S-fl-keyword:modifiers . t)
     (ess-S-fl-keyword:fun-defs  . t)
     (ess-S-fl-keyword:keywords  . t)
-    (ess-S-fl-keyword:assign-ops        . t)
+    (ess-S-fl-keyword:assign-ops . t)
     (ess-S-fl-keyword:constants . t)
     (ess-fl-keyword:fun-calls)
     (ess-fl-keyword:numbers)
     (ess-fl-keyword:operators)
     (ess-fl-keyword:delimiters)
-    (ess-fl-keyword:=)
-    )
+    (ess-fl-keyword:=))
   "An alist of available font-lock keywords for the S mode.
 The key of each cons cell is a name of the keyword. The value
 should be t or nil to indicate if the keyword is activated by
 default or not."
   :group 'ess-S
   :group 'ess-faces
-  :type 'alist)
-
+  :type '(repeat (cons symbol boolean)))
 
-;;; fl-keywords R
 (defvar ess-R-fl-keyword:modifiers
-  '(eval . (cons (concat "\\(" (regexp-opt ess-R-modifyiers 'words) "\\)\\s-*(")
+  '(eval . (cons (concat "\\(" (regexp-opt ess-R-modifiers 'words) "\\)\\s-*(")
                  '(1 ess-modifiers-face)))
-  "Font-lock keyword R modifiers.")
+  "Font-lock keyword R modifiers.
+See `ess-R-modifiers' for the list of modifiers.")
 
 (defvar ess-R-fl-keyword:fun-defs
   '(eval . (cons ess-R-function-name-regexp
                  '(1 font-lock-function-name-face nil)))
-  "Font-lock keyword - function defintions for R.")
+  "Font-lock keyword for function names in function definitions.
+When this keyword is on, function names on the left hand side of
+<- are highlighted with `font-lock-function-name-face'.")
 
-(defvar ess-r--bare-keywords
+(defvar ess-r--non-fn-kwds
   '("in" "else" "break" "next" "repeat"))
 
-(defvar ess-R-fl-keyword:bare-keywords
-  '(eval . (let ((bare-keywords (delq nil (mapcar (lambda (keyword)
-                                                    (car (member keyword ess-r--bare-keywords)))
-                                                  ess-R-keywords))))
-             (cons (regexp-opt bare-keywords 'words) 'ess-keyword-face)))
-  "Font-lock keywords that do not precede an opening parenthesis.")
+(defvar-local ess-r--keyword-regexp nil)
+(defun ess-r-find-fl-keyword (limit)
```

Comment:
I think this should be a private `ess-r--` function.

Reply:
It should indeed.

Reply:
IIUC, changing the keywords variables now takes effect only once per session rather than each time a new buffer is opened? I think we should aim for an effectful configuration UI everywhere we can. Maybe make `ess-r--keyword-regexp` a buffer-local variable?

Reply:
oh it is buffer-local!